rfcs/0000-safe-transmute.md

```diff
@@ -374,88 +374,90 @@ let p = Point { x: 4, y: 2 };
 ```rust
 pub mod crate_a {
 
     #[repr(C)]
-    pub struct Constrained(pub private::ConstrainedInner);
-
-    impl Constrained {
-        /// the sum of `wizz` and `bang` must be greater than or equal to zero.
-        pub fn new(wizz: i8, bang: u8) -> Self {
-            assert!((wizz as i16) + (bang as i16) >= 0);
-            Constrained(private::ConstrainedInner { wizz, bang })
+    pub struct NonEmptySlice<T>(pub private::NonEmptySliceInner<T>);
+
+    impl<T> NonEmptySlice<T> {
+        pub fn from_array<const N: usize>(arr: &[T; N], len: usize) -> Self {
+            assert!(len <= N && len > 0);
+            Self(
+                private::NonEmptySliceInner {
+                    data: arr as *const T,
+                    len,
+                }
+            )
         }
 
-        pub fn something_dangerous(&self) {
-            unsafe {
-                // do something that's only safe if `self.0.wizz + self.0.bang >= 0`
-            }
+        pub fn first(&self) -> &T {
+            unsafe { &*self.0.data }
         }
     }
 
     // introduce a private module to avoid `private_in_public` error (E0446):
     pub(crate) mod private {
-        /// the sum of `wizz` and `bang` MUST be greater than or equal to zero.
         #[repr(C)]
-        pub struct ConstrainedInner {
-            pub wizz: i8,
-            pub bang: u8,
+        pub struct NonEmptySliceInner<T> {
+            pub data: *const T,
+            pub len: usize,
         }
     }
 
 }
 ```
-In the above example, the definitions of both `Constrained` and its field `ConstrainedInner` are marked `pub`. All fields of both types are marked `pub`. However, `Constrained` is *not* fully implicitly constructible outside of `crate_a`, because the module containing `ConstrainedInner` is not visibile outside of `crate_a`.
+In the above example, the definitions of both `NonEmptySlice` and its field `NonEmptySliceInner` are marked `pub`, and all fields of these types are marked `pub`. However, `NonEmptySlice` is *not* fully implicitly constructible outside of `crate_a`, because the module containing `NonEmptySliceInner` is not visibile outside of `crate_a`.
 
 #### Constructability and Transmutation
 Transmutation supplies a mechanism for constructing instances of a type *without* invoking its implicit constructor, nor any constructors defined by the type's author.
 
-In the previous examples, it would be *unsafe* to transmute `0xFF00u16` into `Constrained` outside `crate_a`, because subsequent *safe* use of that value (namely, calling `something_dangerous`) could violate memory safety. (However, it's completely safe to transmute `Constrained` into `0xFF00u16`.)
-
+In the previous examples, it would be *unsafe* to transmute `0u128` into `NonEmptySlice` outside `crate_a`, because subsequent *safe* use of that value (namely, calling `first`) would memory safety. (However, it's completely safe to transmute `NonEmptySlice` into a `u128`.)
```

Comment:
" would memory safety." -> " would violate memory safety."

Reply:
```suggestion
In the previous examples, it would be *unsafe* to transmute `0u128` into `NonEmptySlice` outside `crate_a`, because subsequent *safe* use of that value (namely, calling `first`) would violate memory safety. (However, it's completely safe to transmute `NonEmptySlice` into a `u128`.)
```